connector/src/main/java/com/google/cloud/bigquery/connector/common/StreamCombiningIterator.java

```diff
@@ -0,0 +1,308 @@
+package com.google.cloud.bigquery.connector.common;
+
+import com.google.api.gax.rpc.ResponseObserver;
+import com.google.api.gax.rpc.StreamController;
+import com.google.cloud.bigquery.storage.v1.BigQueryReadClient;
+import com.google.cloud.bigquery.storage.v1.ReadRowsRequest;
+import com.google.cloud.bigquery.storage.v1.ReadRowsResponse;
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.stream.Collectors;
+
+/**
+ * An iterator that combines one or more ReadRows requests into a single iterator. Ordering of
+ * messages is not guaranteed from the streams.
+ *
+ * <p>This is adapated fro ServerStream in gax. In comparison toServerStream, this implementation
+ * also buffers more a configured number of responses instead of a single one. For connections with
+ * high latency between client and server this can processing costs. It also allows combining one or
+ * more ReadRows calls into single iterator to potentially increase perceived client throughput if
+ * that becomes a bottleneck for processing.
+ */
+public class StreamCombiningIterator implements Iterator<ReadRowsResponse> {
+  private static final Object EOS = new Object();
+  // Contains either a ReadRowsResponse, or a terminal object of throwable OR EOS
+  // This class is engineered keep responses below capacity unless a terminal state has
+  // been reached.
+  private final ArrayBlockingQueue<Object> responses;
+  private final ArrayBlockingQueue<Observer> observersQueue;
+  private final AtomicInteger observersLeft;
+  private final int bufferPerStream;
+  private final int numRetries;
+  private final Object lock = new Object();
+  private final BigQueryReadClient client;
+  Object last;
+  volatile boolean cancelled = false;
+  private final Collection<Observer> observers;
+
+  StreamCombiningIterator(
+      BigQueryReadClient client,
+      Collection<ReadRowsRequest.Builder> requests,
+      int bufferPerStream,
+      int numRetries) {
+    this.client = client;
+    observersLeft = new AtomicInteger(requests.size());
+    this.bufferPerStream = bufferPerStream;
+    // + 1 to leave space for terminal object.
+    if (bufferPerStream < 1) {
```

Comment:
Is this value update should be before setting the field (line 50) or after? if after, can you please create another local variable with a different name so it will be clear that the field and the queue sizes may refer to two different values?

Reply:
will change this to not adjust but check the value.

Reply:
done.